tools/generator/GenBaseSupport.cs

```diff
@@ -70,11 +70,8 @@ public ManagedGenBaseSupport (TypeDefinition t)
 			java_name = SymbolTable.FilterPrimitiveFullName (t.FullNameCorrected ());
 			if (java_name == null) {
 				java_name = idx < 0 ? jn : jn.Substring (idx + 1);
-				full_name = t.FullNameCorrected ();
```

Comment:
I'm not sure what to make of this change: semantics for when `java_name != null` changes, removing a `SymbolTable.Lookup(java_name)` invocation. What are the ramifications of this change?

Reply:
From what I was seeing while debugging/testing:
- This line is in the `ManagedGenBaseSupport` class, for C# types
- `full_name` appears to be the full "managed" type name
- The code used to take the `java_name`, and lookup the managed type name from the `SymbolTable`
- Why can't we just use the `TypeDefinition` and get the managed type name?

So it seems to me `t.FullNameCorrected ()` will get the same result, and we aren't hitting the `SymbolTable` anymore.

Does that seem right?

Reply:
(Insert crying here. This will be a long comment.)

This is why we need unit tests. :-/  
Fortunately we have history.  
Unfortunately history isn't "all that".

What we "started" with was:

```csharp
java_name = idx < 0 ? jn : jn.Substring (idx + 1);
full_name = t.FullName.Replace ('/', '.');
```

The seemingly relevant commits are:

* https://github.com/xamarin/monodroid/commit/408b798ba5#diff-e42a846d384e3924dea47ea1c56bc643R43
    > - Treat managed enums differently; they are now mapped to int.
    > - managed primitive types are now correctly treated as primitive symbols.
    ```diff
    -			java_name = idx < 0 ? jn : jn.Substring (idx + 1);
    -			full_name = t.FullName.Replace ('/', '.');
    +			java_name = SymbolTable.FilterPrimitiveFullName (t.FullNameCorrected ());
    +			if (java_name == null) {
    +				java_name = idx < 0 ? jn : jn.Substring (idx + 1);
    +				full_name = t.FullNameCorrected ();
    +			}
    +			else
    +				full_name = SymbolTable.Lookup (java_name).FullName;
    ```
* https://github.com/xamarin/monodroid/commit/fd52336858
    > avoid NRE when name lookup fails (it actually does at mscorlib).
    ```diff
    +			} else {
    +				var sym = SymbolTable.Lookup (java_name);
    +				full_name = sym != null ? sym.FullName : t.FullNameCorrected ();
  			}
    -			else
    -				full_name = SymbolTable.Lookup (java_name).FullName;
    +			java_name = java_name.Replace ('$', '.');
    ```

`t.FullName.Replace('/', '.')` is now `t.FullNameCorrected()`, so it *looks* like *this* PR would nearly return us to a pre-monodroid/e42a846d world, with the primary change being when `SymbolTable.FilterPrimitiveFullName (t.FullNameCorrected ())` is *not* `null`.

[That can happen](https://github.com/xamarin/java.interop/blob/2a9ac6a2a5c746540797d28f2450fb9027d488a0/tools/generator/SymbolTable.cs#L151-L178) when `t.FullName` is the managed name of a primitive type such as `System.Int32` and a particular non-primitive type such as `System.String`.

I then turn my attention to `full_name = SymbolTable.Lookup (java_name).FullName` -- which had to be fixed to `null` check the return value! -- and ask: what type handled by `FilterPrimitiveFullName()` will cause `Lookup(string)` to return `null`?

(insert crying here.)

The answer is `System.String`, because [strings are special-cased and never added to `SymbolTable.symbols`](https://github.com/xamarin/java.interop/blob/2a9ac6a2a5c746540797d28f2450fb9027d488a0/tools/generator/SymbolTable.cs#L57). (Why? No idea. Should this perhaps be fixed? Maybe, but I shudder to think at what all that may impact!)

(It would have been nice for monodroid/fd523368 to mention this, as I can imagine *no other type* which would result in throwing a NRE from that codepath!)

We can thus establish that when `java_name` is *not `null`*, the result of `full_name` will be one of:

1. A "primitive" name: when `t` is a `System.Int32`, then `full_name` will be `int`, *not* `System.Int32`.
2. `System.String` (because `SymbolTable.Lookup ("java.lang.String")` returns `null`, so we fallback to `t.FullNameCorrected()`)

With *that* static analysis out of the way, *does the above change matter*?

I don't know.

I'd have thought that `full_name` value being `System.Int32` instead of `int` would result in a change...except *all of our existing tests* in Java.Interop are *not* using `ManagedGenBaseSupport`, they're using `XmlGenBaseSupport`.

In any event, I think we need some `generator` tests which involve referencing existing assemblies, so that `ManagedGenBaseSupport`/etc. is actually hit. The "nearest match" is probably [`xamarin-android/tests/CodeGen-Binding/Xamarin.Android.McwGen-Tests`](https://github.com/xamarin/xamarin-android/tree/master/tests/CodeGen-Binding/Xamarin.Android.McwGen-Tests).

It would be "interesting" for you to take your updated `generator` and use it to build `CodeGen-Binding/Xamarin.Android.McwGen-Tests`, see what happens.

Reply:
In the meantime, do we need *this* particular "refactoring" in *this* PR? Or can we leave it for later? :-)

Reply:
Ok, I put the code back and was able to move things to `OnValidate` where we actually have an instance of `CodeGenerationOptions` and `SymbolTable`. This should be a safer change.